### DISCUSSION TITLE
Algoritma Problemi

### DISCUSSION THREAD

**Discussion by BunyaminAkdeniz** [19/classroom#72] [Anonim]
Merhabalar Hocam. Elimde bir algoritma sorusu var. Belkide alanda bilinen bir problemdir ama nasıl aratacağımı bilemedim.
Soru şu ki:
Elimizde farklı hacimlerde nesneler var. Bir de bu hacimleri yerleştireceğimiz bir oda var. İstediğim şey odanın maksimum doluluk oranını yakalamak. (Örn. Nesneler: 5, 4, 3, 3, 2, Oda hacmi: 10). Oda toplam hacmi, nesnelerin sayısı ve hacimleri her seferinde değişiyor.
Bir kaç kez bunun kodunu yazmam gerekti, hep etrafından döndüm sanırım. Optimize bir çözüm bulduğumu sanmıyorum. Sizin yaklaşımınızı merak ediyorum. Saygılar.

**Comment by roktas** [comment]
Bu bir optimizasyon problemi. Bu problem özelinde biraz "overkill" kaçabilir emin değilim ama, problemi vesile ederek Google'ın [OR-Tools](https://developers.google.com/optimization) kitaplığıyla tanışmanı istiyorum. Bu tür yüksek seviyeli kitaplıkları tanımanın ve öğrenmenin en etkili yolu budur.

Kitaplığın tanıtımında şöyle bir [ifade](https://developers.google.com/optimization/introduction) geçiyor:

> Bin packing: Pack as many objects of various sizes as possible into a fixed number of bins with maximum capacities.

OR-Tools kullanman için C++ bilmen gerekmiyor. Çeşitli diller için "binding"leri var. (1) Yazdığın dile uygun "binding"i seç, örneğin [Python](https://developers.google.com/optimization/introduction/python); (2) OR-Tools API'yi inceleyerek işine yarayacak kısımları bul ve gerçekle.

**Comment by roktas** [comment]
OR-Tools inanılmaz bir kitaplık. Adamlar bununla endüstriyel zorlukta pek çok problemi gümbür gümbür çözüyorlar. Alet çantanızda böyle bir şeyin olmasında çok yarar var.

P.S. Her zaman olduğu gibi @emres üstadımıza da bir ping atalım. Önerimi (overkill tehlikelisi bir yana) değerlendirebilir veya OR-Tools deneyimi varsa nokta atışı yapabilir.

**Reply by roktas** [reply]
Bir kaç ipucu, anahtar terim... Constraint (yani "kısıt"). Bu problemdeki kısıtın nedir? Linear optimization. Doğrusallık nedir? Neleri birer katsayıyla çarpıp toplaman gerekiyor? Amaç bu katsayıları bulmak mı?

**Reply by roktas** [reply]
https://developers.google.com/optimization/pack/knapsack

Sanırım bu link nokta atışı oldu. Fakat unutma, temelde bu da üst seviye bir doğrusal eniyileştirme algoritmasının özel bir probleme uygulanmış hali. Tavsiyem o üst algoritmaları elinden geldiğince kavraman. (Kağıt kalem kullanarak basit, küçük boyutlu bir problemi o algoritmayla elle çözmeye çalışmak gibisi yoktur.) Derin bir çalışmadan bahsetmiyorum, terminolojiyi, neyi nasıl formüle edeceğini bilmeye yetecek kadar belki.

Optimizasyon problemleri üzerinde geniş bir literatürün oluştuğu iyi bilinen bir sahadır. Bazen karşı karşıya kaldığımız problemin ne denli katmanlı ve derin olduğunu anlamayız. Karmaşık bir hastalığı "köyün şifacısı"ndan aldığımız muhtelif iksirler ve otlarla çözmeye çalışmak gibi kafamızdan bize çok akılcı gelen "heuristic" yöntemlerle çözmeye çalışırız. İşte bu noktada Matematik (ve özelde doğrusal eniyileştirme) gibi insanlığın derin mirası gelmeli aklına. "Köyün şifacısı"na değil "Medical Park" gibi tam teşekküllü bir hastaneye başvurmak gereği ortaya çıkar.

**Reply by emres** [reply]
Nostalji oldu biraz, 1990'larin sonuda ITU'deki Lineer Programlama ve Yöneylem Araştırması (Operations Research) derslerine gittim. Yukarida @roktas nokta vurusu bilgileri paylasmis. Benim uzerinde ciddi proje yapmis oldugum bir alan degil ama ilk aklima gelenlerden biri [Gurobi Optimizer](https://en.wikipedia.org/wiki/Gurobi_Optimizer) oldu, vakti zamaninda bir projede kullanma ihtimali gundeme gelmisti, oradan aklimda kalmis (bir de [GAMS](https://www.gams.com/) gundeme gelmisti).

Kisa sure once yine burada bahsettigim sistemlerden birini tekrar belirtmekte fayda var: [Timefold](https://timefold.ai/).

Bu konular acilinca aklima gelen birisini de belirteyim: [Ozan Erdem](http://www.cs.toronto.edu/~ozan/), kendisine suradan da ulasmak mumkun: https://twitter.com/ozanerdem. Mesela "[Picat](http://picat-lang.org/)" gibi bir ilginc sistemden Ozan sayesinde haberdar olmustum.

Aklima gelen bir baska yazilim da,  Brian W. Kerninghan sayesinde haberdar oldugum [AMPL](https://ampl.com/learn/ampl-book/) (A Modeling Language for Mathematical Programming).

Julia ile gelistirilmis [JuMP](https://jump.dev/JuMP.jl/stable/should_i_use/) var bir de ve onlara dair belki @denizyuret veya @jbytecode daha saglikli bilgi verebilir.

**Reply by Parmendies** [reply]
Çok teşekkür ederim hocam. OR-Tools tam aradığım şey. Oradaki başlıkları ve algoritmaları çalışma listeme ekledim. Hatta bazılarına göz attım ama arka taraf çok korkunç görünüyor... Şimdilik hazıra konucam. Dart için bir kütüphane bulamadım. IPC ile haberleştirmemde bir sakınca var mıdır?

**Reply by roktas** [reply]
@Parmendies Dart binding'i yok gözüküyor: https://github.com/google/or-tools/issues/2830

IPC olur, özelde gRPC sanki daha iyi olur, bk. https://github.com/google/or-tools/issues/953#issuecomment-871728339
